### ISSUE TITLE
Performance issue in StringUtil

### ISSUE THREAD

**Issue by utchibutchi** [jhy/jsoup#1352]
Hi, we're having performance issues with the jsoup lib. We are working on servers with very many cores and want to parse html highly parallel. Unfortunatelly there is a border where we can not increase throughput anymore by increasing threads.

I created a thread dump and nearly all threads are waiting here:

``
INFO   | jvm 1    | 2020/04/17 12:38:23 | "thread-55" ...  waiting for monitor entry
INFO   | jvm 1    | 2020/04/17 12:38:23 |    java.lang.Thread.State: BLOCKED (on object monitor)
INFO   | jvm 1    | 2020/04/17 12:38:23 |       at org.jsoup.internal.StringUtil.borrowBuilder(StringUtil.java:243)
INFO   | jvm 1    | 2020/04/17 12:38:23 |       - waiting to lock <0x00007ef4a2361248> (a java.util.Stack)
INFO   | jvm 1    | 2020/04/17 12:38:23 |       at org.jsoup.internal.StringUtil.normaliseWhitespace(StringUtil.java:146)
INFO   | jvm 1    | 2020/04/17 12:38:23 |       at org.jsoup.nodes.TextNode.text(TextNode.java:46)
``

**Comment by WinstonHuTiger** [comment]
How did you check the thread usage? I mean what tools you used in your project?

**Comment by WinstonHuTiger** [comment]
After we dig into the StringUtil class, we found that after increasing the size of memory, the performance will increase while the number of threads increases.

**Comment by utchibutchi** [comment]
Yes, that's true until a certain point it get's better. But as said, we have very big machines, in detail they have 80 cpu cores, 196GB Memory and a threadpool with the size of up to 200 was used. In the thread dump we could see that nearly all of them were waiting for the monitor as shown above.

As a workaround we now re-implemented the TextNode.text() funktion like this:
```
String wholeText = node.getWholeText();
StringBuilder sb = new StringBuilder();
StringUtil.appendNormalisedWhitespace(sb, wholeText, false);
return sb.toString();
```

To not use a StringBuilder from the pool and it clearly solved the performance problem and we don't see problems related to the many StringBuilders created.

**Comment by jhy** [comment]
Thanks for the report. This is fixed by #1402.
